pkg/lexer/state.go

```diff
@@ -71,47 +83,58 @@ next:
 
 	switch {
 	case l.ch == '\n':
-		// semicolon insertion
+		// auto semicolon insertion
 		if l.insertSemi {
 			l.emit(token.SEMICOLON)
 			return
 		}
 
+		// ignore all space
 		consumeAllSpace(l)
+
 	case unicode.IsSpace(l.ch):
-		// ignore whitespace
-		consumeSpace(l)
+		if l.insertSemi {
+			// semicolon to be inserted
+			// do not ignore newlines
+			consumeSpace(l)
+		} else {
+			// no semicolon to be inserted
+			// ignore all space runes
+			consumeAllSpace(l)
+		}
 
-	// literals
 	case isIdentStart(l.ch):
-		// identifier
 		t := lexIdent(l)
 		l.insertSemi = t.InsertSemi()
+
 	case unicode.IsDigit(l.ch):
-		// number
 		lexNum(l)
+		// semicolon should be inserted after a number
 		l.insertSemi = true
+
 	case l.ch == '"':
-		// format string
 		lexString(l)
+		// semicolon should be inserted after a string
 		l.insertSemi = true
 
-	// operators
+	// all operator starting runes are themselves operators
 	case token.IsOperator(string(l.ch)):
 		t := lexStmtOp(l)
 		l.insertSemi = t.InsertSemi()
 
-	// special
 	case l.ch == '#':
 		// line comment
 		lexComment(l)
+
 	case l.ch == eof:
-		return
+		return // will be handled by lexBase
+
 	default:
-		// rune not supported
+		// rune not supported inside statements
 		l.emit(token.ILLEGAL)
 	}
 
+	// loop till eof
 	goto next
```

Comment:
This has the potential to go disastrous 😂 imagine what would happen if by chance someone has removed the `next:` label from this function...